File: CS/0654_Maximum_Binary_Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode ConstructMaximumBinaryTree(int[] nums) {
        
        if(nums.Length == 0){
            return null;
        }
        
        int targetIndex = 0;
        for(int i = 0; i < nums.Length; i++){
            if(nums[i] > nums[targetIndex]){
                targetIndex = i;
            }
        }
        int[] left = new int[targetIndex];
        int[] right = new int[nums.Length - targetIndex - 1];
        Array.Copy(nums, 0, left, 0, targetIndex);
        Array.Copy(nums, targetIndex+1, right, 0, nums.Length - targetIndex - 1);
        return new TreeNode(nums[targetIndex], ConstructMaximumBinaryTree(left), ConstructMaximumBinaryTree(right));
    }
}

File: Java/0654_Maximum_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if(nums.length == 0){
            return null;
        }
        int maxIndex = 0;
        for(int i = 0; i < nums.length; i++){
            if(nums[i] > nums[maxIndex]){
                maxIndex = i;
            }
        }
        
        int[] leftArray = new int[maxIndex];
        System.arraycopy(nums, 0, leftArray, 0, maxIndex);
        
        int[] rightArray = new int[nums.length - maxIndex-1];
        System.arraycopy(nums, maxIndex+1, rightArray, 0, nums.length - maxIndex-1);
        
        TreeNode result = new TreeNode(nums[maxIndex], constructMaximumBinaryTree(leftArray), constructMaximumBinaryTree(rightArray));
        
        return result;
    }
}
